runtime/codert_vm/xnathelp.m4

```diff
@@ -967,4 +967,23 @@ END_PROC(jitDecompileOnReturnJ)
 
 })	dnl ASM_J9VM_ENV_DATA64
 
+START_PROC(jitReadBarrier)
+	SWITCH_TO_C_STACK
+	SAVE_ALL_REGS
+
+	dnl currentThread->javaVM->memoryManagerFunctions->J9ReadBarrier(currentThread, (fj9object_t*)currentThread->floatTemp1);
+
+	mov    rax,QWORD PTR [rbp+J9TR_VMThread_javaVM]
+	mov    rsi,QWORD PTR [rbp+J9TR_VMThread_floatTemp1]
+	dnl javaVM->memoryManagerFunctions
+	mov    rax,QWORD PTR [rax+0x140]
+	mov    rdi, rbp
+	dnl memoryManagerFunctions->J9ReadBarrier
+	call   QWORD PTR [rax+0x1f8]
+
+	RESTORE_ALL_REGS
+	SWITCH_TO_JAVA_STACK
+	ret
```

Comment:
Yes , it would be nice to use consistent spacing and capitalization, as well as using the const[reg] format rather than [reg+const]. Also, "uword ptr" is preferred for pointer-sized operations (not that I think this code will ever be used on 32-bit), and the _reg syntax for fullword registers.

Also, should this new code be under the 64-bit ifdef?

Reply:
Thank-you for pointing this out, this part is not mature yet, the format will be updated according to both comments.

Reply:
Also, would it be more efficient to either:

- use a cnathelp.cpp helper with an inline (or mostly inline) version of the barrier to avoid the cross-module call?

- call the MM function directly from the compiled code

Another point - this clearly isn't a GC point, so there's no need to save all the registers - you can get away with only saving the C volatile registers (`SAVE_C_VOLATILE_REGS` / `RESTORE_C_VOLATILE_REGS`)

Reply:
Good suggestions. Will definitely evaluate both approaches.
I thought read barrier was a GC point thus I saved all registers. Will update it to only save volatile ones.

Reply:
If it is a GC point, then you would need to build a resolve frame (in cnathelp) before calling the function. I don't see anything similar on Z - wonder what they're doing?

Reply:
Good point.
@fjeremic Are read barriers considered as GC points on Z?

Reply:
From a scan of the J9ReadBarrier implementation, it is certainly not a GC point.

Reply:
The path on Z is quite inefficient - too many registers saved, multiple cross-library calls. There really shouldn't be wildly different paths for X and Z, so I suspect a cleanup of the existing code along with your changes would benefit both architectures.

Reply:
@amicic @dmitripivkine I think this would benefit from an inline .hpp version of the barrier, calling to the GC only when necessary, which could be used in cnathelp.cpp.

Reply:
In ObjectAccessBarrierAPI.hpp there are already APIs like internalPreReadObject/preMixedObjectReadObject/preIndexableObjectReadObject (there is a static variant too, currently not active, but will be eventually be).

However, they end up doing a cross module call. Yes we could inline the implementation to save on cross module (possible one more virtual dispatch and/or another minor check or two). Still, I would not expect much - the barrier itself is relatively heavy.

Reply:
I suppose it also depends on how much of the barrier work is inlined in the compiled code - maybe the call is only for unexpected cases.

Reply:
Fairly little is inlined in compiled code - it's the range check. this is what I meant by 'minor checks' that could be avoided in the OABAPI implementation.

If JIT called the barrier that already means there is non-trivial work to do. as a min (if object is already copied) it will be a ref. update (which is an atomic), or worse to win the copying (atomic) and  copy the object. there are other scenarios, but these two are most common, split about 50-50.

The best savings would come from speeding up the range checks, but this is where only specialized h/w could further help.

Reply:
@gacholio I've updated this method, do you mind take a look once more?

Reply:
> use a cnathelp.cpp helper with an inline (or mostly inline) version of the barrier to avoid the cross-module call?

From the discussions above, the only thing JIT should be inlined is the range check, which is already inlined be JIT. Hence creating a cnathelp.cpp helper is not necessary.

> call the MM function directly from the compiled code

Calling read barrier is considered rare, calling the MM function from JITted code would create too much register pressure and uses too much unnecessary code cache. The whole purpose of this xnathelp.m4 wrapper is to reduce register pressure and save JIT code cache.

Reply:
The spacing is still different. The registers loads are linux-specific, whereas this file is shared with windows as well - might be worth defining a few arg register macros to make this common. All but the first comment seem redundant.

Reply:
Updated. Could you take a look once more?